tracer/src/Datadog.Trace/Span.cs

```diff
@@ -545,5 +548,19 @@ private void WriteCloseDebugMessage()
                 "Span closed: [s_id: {SpanId}, p_id: {ParentId}, t_id: {TraceId}] for (Service: {ServiceName}, Resource: {ResourceName}, Operation: {OperationName}, Tags: [{Tags}])\nDetails:{ToString}",
                 new object[] { Context.RawSpanId, Context.ParentIdInternal, Context.RawTraceId, ServiceName, ResourceName, OperationName, Tags, ToString() });
         }
+
+        internal SpanLink AddSpanLink(Span spanLinkToAdd, List<KeyValuePair<string, string>> attributes = null)
+        {
+            if (IsFinished)
+            {
+                Log.Warning("AddSpanLink should not be called after the span was closed");
+                return null;
```

Comment:
I think the `return null` is a tad risky in a fluent-like interface that `Span.AddSpanLink` and `SpanLink.AddAttribute` have.

I'd see the method signature and think that I should be able to do the following with a fluent interface like this ->

`span.AddSpanLink(someOtherSpan, someAttributes).AddAttribute(someKey, someValue).AddAttribute(someKey2, someValue2);`

but if `span` is marked as `IsFinished` I'll get a `null` everywhere so I have to guard everything with `?.` ->

`span.AddSpanLink(someOtherSpan, someAttributes)?.AddAttribute(someKey, someValue)?.AddAttribute(someKey2, someValue2);`


I'm wondering if changing the return type back to `void` would be the best now for both `AddSpanLink` and `AddAttribute` now that they have that `IsFinished` call.

Reply:
So `Span.SetTag` will always return the `Span` instance and never `null` that's the primary difference here.

For `SetTag`, when `IsFinished` is `true` the tag isn't added, but it will still return the `this` instance of the `Span` compared to the `SpanLink` API returning `null`.

Reply:
For `AddAttribute` maybe I can just return the spanlink then, but for `AddSpanLink` it might be a bit gross. 

I still want to keep the chain-ability of the methods but I'm not sure if it's worth it if it causes an unpleasant experience of using it.

Maybe I can generate the spanlink first, then do the return on the spanlink if the span is closed. 

```
        internal SpanLink AddSpanLink(Span spanLinkToAdd, List<KeyValuePair<string, string>> attributes = null)
        {
            var spanLink = new SpanLink(spanLinkToAdd, this, attributes);
            if (IsFinished)
            {
                Log.Warning("AddSpanLink should not be called after the span was closed");
                return spanLink;
            }

            SpanLinks ??= new List<SpanLink>();
            SpanLinks.Add(spanLink);
            return spanLink;
        }
```

i don't see how this would go wrong - the return would just be a spanlink object but as a user it wouldn't make sense to me that the method failing and succeeding would have the same return.
@lucaspimentel - thoughts?

Reply:
I'm good with the changes 👍